watertap/tools/oli_api/flash.py

```diff
@@ -216,279 +204,346 @@ def _set_prescaling_calculation_mode(self, use_scaling_rigorous):
+        conc_vals = {k: v[s] for k, v in conc_data["values"].items()}
+        inflows.update({"inflows": {"values": conc_vals, "unit": conc_data["unit"]}})
+        if file_name:
+            self.write_output(inflows, file_name)
+        return inflows
 
-            def _filter_zeroes(data):
-                if "values" in data:
-                    data["values"] = {k: v for k, v in data["values"].items() if v != 0}
-                return data
+    def extract_properties(
+        self, raw_result, properties, samples=None, filter_zero=True, file_name=""
+    ):
+        """
+        Extract specified properties from OLI Cloud stream output JSON.
 
+        :param raw_result: dict OR string of JSON file to extract from
+        :param properties: list of properties to extract
+        :param samples: list of indices to extract
+        :param phase: string for phase key of property
+        :param filter_zero: bool switch to exclude zero values properties
+        :param file_name: string for file to write, if any
+
+        :return extracted_properties: dictionary of specified properties and values
+        """
+
+        def _get_nested_paths(data, prop, path=[], nested_paths=[]):
+            """
+            Get nested paths for all occurences of properties in OLI result.
+
+            :param data: dictionary of nested data
+            :param prop: string for property name
+            :param path: empty list to initialize path search
+            :param nested_paths: list for tracking paths
+
+            :return nested_paths: list for tracking paths
+            """
+
+            for k, v in data.items():
+                path.append(k)
+                if hasattr(v, "items"):
+                    if "unit" not in v:
+                        _get_nested_paths(v, prop)
+                if k == prop:
+                    nested_paths.append(deepcopy(path))
+                del path[-1]
+            return nested_paths
+
+        def _get_filtered_result(data, keys, samples, filter_zero):
+            """
+            Get unit and values from nested dictionary data.
+
+            :param data: dictionary of nested data
+            :param keys: list of keys for property
+            :param samples: list of indices to extract
+            :param filter_zero: bool switch to exclude zero values properties
+
+            :return filtered_result: dictionary for unit and values
+            """
+
+            # get nested data
+            for key in keys:
+                data = data[key]
+            nested_data = data
+            unit = nested_data["unit"]
+            # sample nested data
+            if "value" in data:
+                sampled_data = [nested_data["value"][s] for s in samples]
+            elif "values" in nested_data:
+                sampled_data = {
+                    k: [v[s] for s in samples] for k, v in nested_data["values"].items()
+                }
```

Comment:
No `else` -- `sampled_data` won't exist below.

Should catch the `else`.
```suggestion
                }
            else:
                raise RuntimeError("a meaningful message")
```